subsys/bluetooth/host/hci_core.c

```diff
@@ -536,41 +537,85 @@ static int le_set_private_addr(u8_t id)
 	return err;
 }
 
-static void rpa_timeout(struct k_work *work)
+static void le_update_private_addr(void)
 {
-	int err_adv = 0, err_scan = 0;
-
-	BT_DBG("");
-
-	/* Invalidate RPA */
-	atomic_clear_bit(bt_dev.flags, BT_DEV_RPA_VALID);
+	bool adv_enabled = false;
+	int err;
 
 	/*
 	 * we need to update rpa only if advertising is ongoing, with
 	 * BT_DEV_KEEP_ADVERTISING flag is handled in disconnected event
 	 */
 	if (atomic_test_bit(bt_dev.flags, BT_DEV_ADVERTISING)) {
-		/* make sure new address is used */
 		set_advertise_enable(false);
-		err_adv = le_set_private_addr(bt_dev.adv_id);
-		set_advertise_enable(true);
+		adv_enabled = true;
 	}
 
-	if (atomic_test_bit(bt_dev.flags, BT_DEV_ACTIVE_SCAN)) {
-		/* TODO do we need to toggle scan? */
-		err_scan = le_set_private_addr(BT_ID_DEFAULT);
+#if defined(CONFIG_BT_OBSERVER)
+	bool scan_enabled = false;
+
+	if (atomic_test_bit(bt_dev.flags, BT_DEV_SCANNING) &&
+	    atomic_test_bit(bt_dev.flags, BT_DEV_ACTIVE_SCAN)) {
+		set_le_scan_enable(BT_HCI_LE_SCAN_DISABLE);
+		scan_enabled = true;
+	}
+#endif
```

Comment:
Could this be put behind `IS_ENABLED()`? You'd then need to move `set_le_scan_enable()` away from #ifdef (which in this case can be done since it doesn't look like it depends on conditional struct members or anything like that)

Reply:
It could. Will include this as a refactor commit at the end. Since this pattern is present in code not part of this PR. (If needed I can include refactor at the beginning and amend this commit).

Reply:
Unfortunately this requires more changes since all functions that call `le_set_scan_enabl()` is under some form of ifdef, and fixing them causes new dependencies.
Either SMP, PRIVACY, CONN, CENTRAL, OBSERVER will cause problems. Which means non-connectable advertiser will cause an `defined but not reference` warning here.

Removing this change again. Must be done as part of a larger refactor.